src/main/java/Astar.java

```diff
@@ -27,29 +27,27 @@ public Astar(FifteenPuzzle currentBoard, String heuristic, int rows, int columns
     }
 
     public void solveHamm() {
-        if (stats.checkAndSaveStats(this.currentBoard)) return;
-        try {
-            move();
-        } catch (CloneNotSupportedException e) {
-            e.printStackTrace();
-        }
-        this.currentBoard = boardsToCheck.remove(0);
-        if (this.boardsToCheck.size() != 0) {
-            this.solveHamm();
-        }
+        do {
+            if (stats.checkAndSaveStats(this.currentBoard)) return;
+            try {
+                move();
+            } catch (CloneNotSupportedException e) {
+                e.printStackTrace();
+            }
+            this.currentBoard = boardsToCheck.remove(0);
+        } while (this.boardsToCheck.size() != 0);
     }
 
     public void solveManh() {
-        if (stats.checkAndSaveStats(this.currentBoard)) return;
-        try {
-            move();
-        } catch (CloneNotSupportedException e) {
-            e.printStackTrace();
-        }
-        this.currentBoard = boardsToCheck.remove(0);
-        if (this.boardsToCheck.size() != 0) {
-            this.solveManh();
-        }
+        do {
+            if (stats.checkAndSaveStats(this.currentBoard)) return;
+            try {
+                move();
+            } catch (CloneNotSupportedException e) {
+                e.printStackTrace();
+            }
+            this.currentBoard = boardsToCheck.remove(0);
+        } while (this.boardsToCheck.size() != 0);
     }
```

Comment:
This methods are identical, so it can be only one without switch in constructor.